### PR TITLE
Linreg

### PR DESCRIPTION
**Pull request by aisthesis**
Simple tool for performing regularized linear regression.

### CODE REVIEW

**Comment by aisthesis** [comment]
Here's where I'm coming from on the formula: http://stackoverflow.com/questions/27476933/numpy-linear-regression-with-regularization

**Comment by aisthesis** [comment]
I'm pretty confident that the formula is correct. Here's my result using `sklearn`:

    >>> from sklearn.linear_model import Ridge
    >>> clf = Ridge(alpha=3., solver='svd')
    >>> clf.fit(features['train'], labels['train'])
    Ridge(alpha=3.0, copy_X=True, fit_intercept=True, max_iter=None,
       normalize=False, random_state=None, solver='svd', tol=0.001)
    >>> clf.coef_
    array([ 0.        ,  6.52208421,  3.82487688,  3.61985418,  2.18396316,
            2.10604481,  1.2807778 ,  1.28789997,  0.72492414])
    >>> clf.intercept_
    11.217589325366376
    >>> predicted = clf.predict(features['test'])
    >>> pn.learn.mse(predicted, labels['test'])
    array([[ 7.14405231]])

File: pynance/learn/linreg.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    model : ndarray
+        Regression model for the given data.
+    """
+    n_col = (features.shape[1] if len(features.shape) > 1 else 1)
+    reg_matrix = regularization * np.identity(n_col, dtype='float64')
+    if constfeat:
+        reg_matrix[0, 0] = 0.
+    return np.linalg.lstsq(features.T.dot(features) + reg_matrix, features.T.dot(labels))[0]
+
+def predict(features, model):
```
Comment (aisthesis) [inline pynance/learn/linreg.py]: Test on toy data.

File: pynance/learn/metrics.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+Metrics (:mod:`pynance.learn.metrics`)
+===============================================
+
+.. currentmodule:: pynance.learn.metrics
+"""
+
+import numpy as np
+
+def mse(predicted, actual):
```
Comment (aisthesis) [inline pynance/learn/metrics.py]: test on toy data

File: pynance/learn/metrics.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    -------
+    err : ndarray
+        Mean squared error of predictions relative to actual
+        values.
+    """
+    diff = predicted - actual
+    ncols = (diff.shape[1] if len(diff.shape) > 1 else 1) 
+    return np.average(diff * diff, axis=0).reshape(1, ncols)
+
+def stderr(predicted, actual):
```
Comment (aisthesis) [inline pynance/learn/metrics.py]: test on toy data

File: pynance/learn/linreg.py (hunk 1 above)
Comment (arkadasgupta) [inline pynance/learn/linreg.py]: add a link to the SO post.

File: pynance/learn/linreg.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    Returns
+    -------
+    model : ndarray
+        Regression model for the given data.
+    """
+    n_col = (features.shape[1] if len(features.shape) > 1 else 1)
+    reg_matrix = regularization * np.identity(n_col, dtype='float64')
+    if constfeat:
+        reg_matrix[0, 0] = 0.
+    return np.linalg.lstsq(features.T.dot(features) + reg_matrix, features.T.dot(labels))[0]
```
Comment (arkadasgupta) [inline pynance/learn/linreg.py]: add a link to the SO post

File: pynance/learn/metrics.py (hunk 3 above)
Comment (arkadasgupta) [inline pynance/learn/metrics.py]: instead of stderr, name it 'rmse' ?

File: pynance/tst/unit/learn/prep_jwht.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    ifname = 'Hitters.csv'
+    infile = os.path.join(path, ifname)
+    src_data = read_src(infile)
+    convert_nonnum(src_data)
+    features, labels = get_labeledfeatures(src_data)
+    ofname = 'hitters.data'
+    outfile = os.path.join(path, ofname)
+    save_data(features, labels, outfile)
+
+if __name__ == '__main__':
```
Comment (arkadasgupta) [inline pynance/tst/unit/learn/prep_jwht.py]: i would prefer this to be somewhere other than 'learn'. maybe in 'data'

File: pynance/tst/unit/learn/test_linreg.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    features = pd.DataFrame(data=data.values.copy(), 
+            columns=['Constant', 'Out', 'Old', 'Pop'])
+    features.iloc[:, 0] = 1.
+    labels = pd.DataFrame(data=data.values[:, 0])
+    return features, labels
+
+def adj_yule(df):
+    return df - 100.
+
+def load_yule():
```
Comment (arkadasgupta) [inline pynance/tst/unit/learn/test_linreg.py]: abstract it out of the test

File: pynance/tst/unit/learn/test_linreg.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    path = os.path.dirname(os.path.realpath(__file__))
+    infile = os.path.join(path, 'yule.data')
+    data = pd.read_table(infile, sep=' ', index_col=False, header=None,
+            names=['Paup', 'Out', 'Old', 'Pop'], dtype=np.float64)
+    return data
+
+def get_ng_error(features, labels, regul, key):
```
Comment (arkadasgupta) [inline pynance/tst/unit/learn/test_linreg.py]: i don't think putting ng in the function name to indicate it's from Ng's data is a good idea. ng has a lot of different meanings in a lot of different contexts. Maybe coursera is better

File: pynance/tst/unit/learn/x.data [hunk 8]
```diff
@@ -0,0 +1,12 @@
+ -15.93675813378541
```
Comment (arkadasgupta) [inline pynance/tst/unit/learn/x.data]: i'd put the data files in a different directory.

**Comment by arkadasgupta** [comment]
i mostly have some minor comments. other than that it looks good to go

File: pynance/learn/metrics.py (hunk 3 above)
Comment (aisthesis) [inline pynance/learn/metrics.py]: 'standard error' is the common usage in statistics.

File: pynance/learn/linreg.py (hunk 4 above)
Comment (aisthesis) [inline pynance/learn/linreg.py]: I don't think the documentation should include implementation details.

File: pynance/tst/unit/learn/test_linreg.py (hunk 6 above)
Comment (aisthesis) [inline pynance/tst/unit/learn/test_linreg.py]: not sure what you mean

File: pynance/tst/unit/learn/test_linreg.py (hunk 7 above)
Comment (aisthesis) [inline pynance/tst/unit/learn/test_linreg.py]: coursera has numerous courses on machine learning, and I don't think we should get too hung up on naming for testing functions. I just want to give a nod to where I'm getting it.

File: pynance/tst/unit/learn/x.data (hunk 8 above)
Comment (aisthesis) [inline pynance/tst/unit/learn/x.data]: I considered that. If we end up with a lot of tests that rely on data sets that need to be read from files, then I think we should do that. But I don't expect that to be the case. Usually, we can just use toy data, as in your integration test. Within the directory, it's easy to distinguish between actual test files (which have to begin with 'test' for nose to run them), other python files (ending in `.py` but not starting with `test`) and data files (labeled `data`).

Currently, this is the only directory that even *has* this kind of structure, and it's very easy to see which files are which in light of naming. Moving to a different directory would also make the coding slightly more convoluted in that the paths wouldn't be local. I agree with you if this kind of things grows more but feel like it makes things simpler for now to put the data with the test that draws from it. Moreover, factoring out a separate directory suggests that it might be data used by more than 1 test. That isn't the case, and it's doubtful it ever will be.

File: pynance/learn/linreg.py (hunk 4 above)
Comment (arkadasgupta) [inline pynance/learn/linreg.py]: no i meant as a line comment.

File: pynance/tst/unit/learn/test_linreg.py (hunk 6 above)
Comment (arkadasgupta) [inline pynance/tst/unit/learn/test_linreg.py]: i meant use a sort of util file for these things

**Comment by arkadasgupta** [comment]
some replies on your comments. other than that good to go.
